### PR TITLE
Proper nrunner message saving

### PR DESCRIPTION
**Pull request by richtja**
The saving of the nrunner messages was too strict. The stdout and stderr
messages could be only bytes and the log messages could be only stings
without a new line at the end. This commit makes the saving more
flexible.

Signed-off-by: Jan Richter <jarichte@redhat.com>

### CODE REVIEW

File: avocado/core/messages.py [hunk 1]
```diff
@@ -147,10 +147,33 @@ class BaseRunningMessageHandler(BaseMessageHandler):
...
+        It can save str or byte types of messages. When the str message doesn't
+        end with a new line the new line will be added. Every message is saved
+        in the append mode.
+
+        :param filename: name of the file
+        :type filename: str
+        :param buff: message to be saved
+        :type buff: str or bytes
+        :param task: message related task.
+        :type: :class:`avocado.core.nrunner.Task`
```
Comment (clebergnu) [inline avocado/core/messages.py:163]: Missing `task` here for this `:type:` line.

File: avocado/core/messages.py [hunk 2]
```diff
@@ -147,10 +147,33 @@ class BaseRunningMessageHandler(BaseMessageHandler):
     """Base interface for resolving running messages."""
 
     @staticmethod
-    def _save_message_to_file(filename, buff, task, mode='a'):
+    def _save_message_to_file(filename, buff, task):
```
Comment (clebergnu) [inline avocado/core/messages.py:150]: There's a typo in the commit message: "be only **stings**"

File: avocado/core/messages.py [hunk 3]
```diff
@@ -147,10 +147,33 @@ class BaseRunningMessageHandler(BaseMessageHandler):
...
+        :param filename: name of the file
+        :type filename: str
+        :param buff: message to be saved
+        :type buff: str or bytes
+        :param task: message related task.
+        :type: :class:`avocado.core.nrunner.Task`
+        """
+
+        def _save_to_file(mode):
+            with open(file, mode) as fp:
```
Comment (clebergnu) [inline avocado/core/messages.py:167]: I understand that `file` will be evaluated later, but I find it extremely confusing to refer to a variable in a different scope that has not been defined yet (in the code order).

Maybe it's just me, but I'd like to hear from others (@willianrampazzo @beraldoleal).

**Review by clebergnu — CHANGES_REQUESTED** [review]
Hi @richtja ,

I found a typo and a missing part of a docstring.  Besides that, an style issue that @willianrampazzo @beraldoleal (and others) may give their opinion too about.

Thanks!

**Comment by willianrampazzo** [comment]
> Hi @richtja ,
> 
> I found a typo and a missing part of a docstring. Besides that, an style issue that @willianrampazzo @beraldoleal (and others) may give their opinion too about.
> 
> Thanks!

I was holding my review because, if I recall correctly, in one of our latest scrum meetings or in a comment from a PR there was a mention to do not change the content that goes to the log file adding new lines or using other than types bytes.

So, what should be the approach of Avocado related to how it is logging the content? Or, did I make confusion with what was mentioned and what this PR proposes?

**Comment by clebergnu** [comment]
> > Hi @richtja ,
> > I found a typo and a missing part of a docstring. Besides that, an style issue that @willianrampazzo @beraldoleal (and others) may give their opinion too about.
> > Thanks!
> 
> I was holding my review because, if I recall correctly, in one of our latest scrum meetings or in a comment from a PR there was a mention to do not change the content that goes to the log file adding new lines or using other than types bytes.

Yes, that's a good point, and that needs to be addressed. I guess I was (mistakenly) considering this a refactor (with improvements such as the docstring). 

> 
> So, what should be the approach of Avocado related to how it is logging the content? Or, did I make confusion with what was mentioned and what this PR proposes?

Building on my previous comment, I guess the issues we have to decide include:

* Should we allow *any* log messages to contain string in addition to bytes?  If so, what's the benefit .vs. complexity?
* Should we allow for any/some/certain log messages to *append* to a file? IMO this should only happen if we formalize/document support for streaming log messages, and runners could have the ability to send the log (including partial `stdout`, `stderr` at any time).
* Should runners be able to *also* send a complete version of log files after having sent the partial data through streaming?

File: avocado/core/messages.py (hunk 3 above)
Comment (beraldoleal) [reply avocado/core/messages.py:167]: Indeed, also `_save_to_file()` seems to me that is waiting for a file to save. I would move the filename as the first argument here.

**Review by beraldoleal — CHANGES_REQUESTED** [review]

File: avocado/core/messages.py (hunk 3 above)
Comment (willianrampazzo) [reply avocado/core/messages.py:167]: It looks weird for me too, but as this is a function inside a method, it has access to its context. Shame on me. I also used this construction before :P

**Comment by richtja** [comment]
> > So, what should be the approach of Avocado related to how it is logging the content? Or, did I make confusion with what was mentioned and what this PR proposes?
> 
> Building on my previous comment, I guess the issues we have to decide include:
> 
>    * Should we allow _any_ log messages to contain string in addition to bytes?  If so, what's the benefit .vs. complexity?

My opinion is that we should leave the decision to the runners and avocodo have to be able to process the message in the same format as it is sent by the runner. Because the creator of the runner should have the power to create the logs as he wants to and how it is suites to his usage. This is in contradiction with my code, and I agree with @willianrampazzo  that addition of a new line is wrong, I will change that.  
> 
>    * Should we allow for any/some/certain log messages to _append_ to a file? IMO this should only happen if we formalize/document support for streaming log messages, and runners could have the ability to send the log (including partial `stdout`, `stderr` at any time).

Maybe I don't understand what you mean by this, because the messages already do it.
> 
>    * Should runners be able to _also_ send a complete version of log files after having sent the partial data through streaming?

When we are using stable communication, I don't think it is a good idea to send the data which was already sent, because it can be confusing, and it will bring more complexity about dealing with duplicates.

**Comment by clebergnu** [comment]
> > > So, what should be the approach of Avocado related to how it is logging the content? Or, did I make confusion with what was mentioned and what this PR proposes?
> > 
> > 
> > Building on my previous comment, I guess the issues we have to decide include:
> > 
> > * Should we allow _any_ log messages to contain string in addition to bytes?  If so, what's the benefit .vs. complexity?
> 
> My opinion is that we should leave the decision to the runners and avocodo have to be able to process the message in the same format as it is sent by the runner. Because the creator of the runner should have the power to create the logs as he wants to and how it is suites to his usage. This is in contradiction with my code, and I agree with @willianrampazzo that addition of a new line is wrong, I will change that.
> 

I think I understood your point, but just to make sure (a Y/N answer is ideal), in your opinion should we allow _any_ log messages (as sent from the runner) to contain string in addition to bytes?

And +1 for removing the new line.

> > * Should we allow for any/some/certain log messages to _append_ to a file? IMO this should only happen if we formalize/document support for streaming log messages, and runners could have the ability to send the log (including partial `stdout`, `stderr` at any time).
> 
> Maybe I don't understand what you mean by this, because the messages already do it.
> 

My question is if every new `log` message, from the sender perspective, should be considered complete or partial.  If it's always going to be partial, then it makes sense to also always append to the resulting file.  I tend to agree with that approach, but we must formalize that behavior, so that people writing runners don't send partial messages, and at the end of a test send the full content again.

> > * Should runners be able to _also_ send a complete version of log files after having sent the partial data through streaming?
> 
> When we are using stable communication, I don't think it is a good idea to send the data which was already sent, because it can be confusing, and it will bring more complexity about dealing with duplicates.

Agreed, but again, it's about formalizing what runners are expected to do.

With all the questions answered, we should be able to come up with logging specification/rules such as:

* All content is treated as binary
* Runners can send the encoding if logs are to be human readable
* All logged content is partial and will be appended to a file
* If a partial content is to be decoded and shown to humans, the optional encoding will be used and if missing, a newline will be added *during presentation* only (not while saving the content to the file)

**Comment by richtja** [comment]
@clebergnu, I agree that we have to formalize the logging specification/rules, but I think that just formalizing the is not enough. We have to use a more descriptive approach which will make the new runner's creation easier. I am talking about some serialization tool like protobuf, which will define  the data and help users with logging.

> With all the questions answered, we should be able to come up with logging specification/rules such as:
> 
>    * All content is treated as binary

I agree with this, because the binary content with optional encoding gives the users power to create the logs as they want.
> 
>    * Runners can send the encoding if logs are to be human readable
> 
>    * All logged content is partial and will be appended to a file

Yes I believe that  logged content should be treated as partial.
> 
>    * If a partial content is to be decoded and shown to humans, the optional encoding will be used and if missing, a newline will be added _during presentation_ only (not while saving the content to the file)

I am not sure what you mean by presentation_ only, because when we save something to 'debug.log' file, it should be human-readable.

**Comment by clebergnu** [comment]
> @clebergnu, I agree that we have to formalize the logging specification/rules, but I think that just formalizing the is not enough. We have to use a more descriptive approach which will make the new runner's creation easier. I am talking about some serialization tool like protobuf, which will define the data and help users with logging.
> 

I'm all in favor of making the creation of new runners easier.  But let me say that I'm skeptical about resorting to an external RPC tool at this point.  I've had mixed (to put it mildly) experiences with CORBA, ICE, SOAP, XML/RPC, JSON/RPC, gRPC and others.

I understand that the problem we have is that messages may be created with structure and content not matching the specification.  IMO, we can solve most of the problem by:

* More helper methods on base classes, that will produce the right content
* Introducing Python type hints, so that wrong content can be flagged
* Introducing some tests doing JSON schema validation 

> > With all the questions answered, we should be able to come up with logging specification/rules such as:
> > 
> > * All content is treated as binary
> 
> I agree with this, because the binary content with optional encoding gives the users power to create the logs as they want.
> 

Great.

> > * Runners can send the encoding if logs are to be human readable
> > * All logged content is partial and will be appended to a file
> 
> Yes I believe that logged content should be treated as partial.
> 
> > * If a partial content is to be decoded and shown to humans, the optional encoding will be used and if missing, a newline will be added _during presentation_ only (not while saving the content to the file)
> 
> I am not sure what you mean by presentation_ only, because when we save something to 'debug.log' file, it should be human-readable.

That's exactly my point.  `debug.log` is *only* supposed to be human-readable, while `stdout` can be said to be human readable too **but** should never get spurious newlines added to them.

**Comment by richtja** [comment]
> > @clebergnu, I agree that we have to formalize the logging specification/rules, but I think that just formalizing the is not enough. We have to use a more descriptive approach which will make the new runner's creation easier. I am talking about some serialization tool like protobuf, which will define the data and help users with logging.
> 
> I'm all in favor of making the creation of new runners easier. But let me say that I'm skeptical about resorting to an external RPC tool at this point. I've had mixed (to put it mildly) experiences with CORBA, ICE, SOAP, XML/RPC, JSON/RPC, gRPC and others.
> 
> I understand that the problem we have is that messages may be created with structure and content not matching the specification. IMO, we can solve most (if not all) of the problems by:
> 
>     * Adding more/better helper methods on base classes, that will produce the right content
> 
>     * Introducing Python type hints, so that wrong content can be flagged
> 
>     * Introducing some tests doing JSON schema validation
> 

I agree with you those tasks will solve the problem, so I created issue #4627 to track them.

> 
> > > With all the questions answered, we should be able to come up with logging specification/rules such as:
> > > 
> > > * All content is treated as binary
> > 
> > 
> > I agree with this, because the binary content with optional encoding gives the users power to create the logs as they want.
> 
> Great.
> 
> > > * Runners can send the encoding if logs are to be human readable
> > > * All logged content is partial and will be appended to a file
> > 
> > 
> > Yes I believe that logged content should be treated as partial.
> > > * If a partial content is to be decoded and shown to humans, the optional encoding will be used and if missing, a newline will be added _during presentation_ only (not while saving the content to the file)
> > 
> > 
> > I am not sure what you mean by presentation_ only, because when we save something to 'debug.log' file, it should be human-readable.
> 
> That's exactly my point. `debug.log` is _only_ supposed to be human-readable, while `stdout` can be said to be human readable too **but** should never get spurious newlines added to them.

Thanks, @clebergnu, for your comments. I will create new version with binary messaging.

**Comment by richtja** [comment]
Closing this in favor of #4629

**Pull request closed without merging**